### PR TITLE
feat: add the @dim syntax

### PR DESCRIPTION
**Pull request by DongHY1**
#2

### CODE REVIEW

File: packages/shaku-code-annotate/src/index.ts [hunk 1]
```diff
@@ -211,23 +237,17 @@ type ShakuComponent = ShakuComponentCallout | ShakuComponentUnderline;
...
   switch (component.type) {
     case "ShakuComponentCallout":
-      return `<div class="shaku-callout" style="left:${
-        component.config.offset
-      }ch"><span class="shaku-callout-arrow" style="left:${
-        component.config.arrowOffset
-      }ch"></span>${component.config.contents
-        .map(escapeHtml)
-        .join("\n")}</div>`;
+      return `<div class="shaku-callout" style="left:${component.config.offset
```
Comment (JSerZANP) [inline packages/shaku-code-annotate/src/index.ts:240]: what is the cause of these changes?

please run `npm run format` in the root dir

File: packages/remark-shaku-code-annotate/src/__tests__/index.test.ts [hunk 2]
```diff
@@ -137,6 +137,6 @@ const c = 1;
...
 `);
 });
```
Comment (JSerZANP) [inline packages/remark-shaku-code-annotate/src/__tests__/index.test.ts:142]: shall we also add a test case for the proposed command here?

File: examples/web/components/Playground.tsx [hunk 3]
```diff
@@ -53,16 +53,20 @@ const blog = "jser.dev"
...
-**easily Highlight lines**
+**easily Highlight/Hidden lines**
 
-You don't need to add code meta, just add the command where you want to highlight
+You don't need to add code meta, just add the command where you want to highlight or hidden.
 
 \`\`\`js annotate
 // @highlight
 function useSomeEffect({blog}) {
+// @hidden
```
[the remark this answers is not in the document]
Comment (DongHY1) [reply examples/web/components/Playground.tsx:63]: > Shall we choose a different command name for this ? we are actually not hiding the lines but deemphasize it I'd like to use `@hidden` similar to `cut` in twoslash.
> 
> How about `deemphasize`?.

It sounds good, but might it be slightly inconvenient to use due to the length of the word? Perhaps considering 'focus' or 'defocus'? [This Project](https://codehike.org/demo/code) implements similar features to ours.

File: packages/remark-shaku-code-annotate/src/__tests__/index.test.ts [hunk 4]
```diff
@@ -101,7 +101,7 @@ const a = 1;
...
-    .toEqual(`<pre class="shiki" style="color:#24292e;background-color:#fff"><div class="code-container"><code><div class="line"><span style="color: #D73A49">const</span><span style="color: #24292E"> </span><span style="color: #005CC5">a</span><span style="color: #24292E"> </span><span style="color: #D73A49">=</span><span style="color: #24292E"> </span><span style="color: #005CC5">1</span><span style="color: #24292E">;</span></div><div class="shaku-underline shaku-underline-wavy" style="left:6ch"><span class="shaku-underline-line" style="left:0ch">~~~~~</span>This is line 1
+    .toEqual(`<pre class="shiki" style="color:#24292e;background-color:#fff"><div class="code-container"><code><div style="opacity:1" class="line"><span style="color: #D73A49">const</span><span style="color: #24292E"> </span><span style="color: #005CC5">a</span><span style="color: #24292E"> </span><span style="color: #D73A49">=</span><span style="color: #24292E"> </span><span style="color: #005CC5">1</span><span style="color: #24292E">;</span></div><div class="shaku-underline shaku-underline-wavy" style="left:6ch"><span class="shaku-underline-line" style="left:0ch">~~~~~</span>This is line 1
```
[the remark this answers is not in the document]
Comment (DongHY1) [reply packages/remark-shaku-code-annotate/src/__tests__/index.test.ts]: fix it

File: packages/remark-shaku-code-annotate/src/index.ts [hunk 5]
```diff
@@ -165,11 +183,11 @@ export const remarkShakuCodeAnnotate = (
...
               isHighlightingBlock || shouldHighlighNextSourceLine;
             shouldHighlighNextSourceLine = false;
-
+            const shouldHidden = isHiddenBlock || shouldHiddenNextSourceLine;
+            shouldHiddenNextSourceLine = false
             const sourceLine = line.line;
-            const prefix = `<div class="line${
-              shouldHighlight ? " highlight" : ""
-            }">`;
+            const prefix = `<div style="opacity:${shouldHidden ? 0.3 : 1}" class="line${shouldHighlight ? " highlight" : ""
```
[the remark this answers is not in the document]
Comment (DongHY1) [reply packages/remark-shaku-code-annotate/src/index.ts]: > Let's not use hardcode opacity in style here, rather let's use class name so people could easily customize through CSS, as explained here https://github.com/JSerZANP/shaku/tree/main/packages/remark-shaku-code-annotate

I removed the hard-coded styles and added a "defocus" class name. Now users can adjust the opacity by customizing CSS.

File: packages/shaku-code-annotate/src/index.ts (hunk 1 above)
Comment (DongHY1) [reply packages/shaku-code-annotate/src/index.ts]: fix it

File: packages/remark-shaku-code-annotate/src/__tests__/index.test.ts (hunk 2 above)
Comment (DongHY1) [reply packages/remark-shaku-code-annotate/src/__tests__/index.test.ts:161]: add test case

File: packages/shaku-code-annotate/src/index.ts (hunk 1 above)
Comment (DongHY1) [reply packages/shaku-code-annotate/src/index.ts]: Forgot to format the code before submitting. Thanks for catching that!

File: packages/remark-shaku-code-annotate/src/__tests__/index.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (DongHY1) [reply packages/remark-shaku-code-annotate/src/__tests__/index.test.ts]: > this `style="opacity:1"` doesn't seem to be necessary, see the comment on packages/remark-shaku-code-annotate/src/index.ts

I removed the hard-coded styles and added a "defocus" class name. Now users can adjust the opacity by customizing CSS.

File: examples/web/components/Playground.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (JSerZANP) [reply examples/web/components/Playground.tsx]: > but might it be slightly inconvenient but might it be slightly inconvenient 

Agreed, let's use `@dim` then.

I think we should separate the usage of these 3

1. `@highlight` => to emphasize some lines
2. `@dim` => to de-emphasize some lines
3. `@focus` => emphasize some lines while de-emphasize the others, could be thought as a shorthand.

File: packages/remark-shaku-code-annotate/src/__tests__/index.test.ts [hunk 6]
```diff
@@ -41,7 +41,7 @@ const a = 1;
 \`\`\`
 `);
   expect(result.value)
-    .toEqual(`<pre class="shiki" style="color:#24292e;background-color:#fff"><div class="code-container"><code><div class="line"><span style="color: #D73A49">const</span><span style="color: #24292E"> </span><span style="color: #005CC5">a</span><span style="color: #24292E"> </span><span style="color: #D73A49">=</span><span style="color: #24292E"> </span><span style="color: #005CC5">1</span><span style="color: #24292E">;</span></div><div class="shaku-callout" style="left:6ch"><span class="shaku-callout-arrow" style="left:0ch"></span>This is line 1
+    .toEqual(`<pre class="shiki" style="color:#24292e;background-color:#fff"><div class="code-container"><code><div class="line "><span style="color: #D73A49">const</span><span style="color: #24292E"> </span><span style="color: #005CC5">a</span><span style="color: #24292E"> </span><span style="color: #D73A49">=</span><span style="color: #24292E"> </span><span style="color: #005CC5">1</span><span style="color: #24292E">;</span></div><div class="shaku-callout" style="left:6ch"><span class="shaku-callout-arrow" style="left:0ch"></span>This is line 1
```
Comment (JSerZANP) [inline packages/remark-shaku-code-annotate/src/__tests__/index.test.ts:44]: please remove the extra space

File: packages/remark-shaku-code-annotate/src/__tests__/index.test.ts [hunk 7]
```diff
@@ -56,7 +56,7 @@ const a = 1;
...
```
Comment (JSerZANP) [inline packages/remark-shaku-code-annotate/src/__tests__/index.test.ts:59]: same as above

File: packages/remark-shaku-code-annotate/src/__tests__/index.test.ts [hunk 8]
```diff
@@ -71,7 +71,7 @@ const a = 1;
...
 `);
   expect(result.value)
-    .toEqual(`<pre class="shiki" style="color:#24292e;background-color:#fff"><div class="code-container"><code><div class="line"><span style="color: #D73A49">const</span><span style="color: #24292E"> </span><span style="color: #005CC5">a</span><span style="color: #24292E"> </span><span style="color: #D73A49">=</span><span style="color: #24292E"> </span><span style="color: #005CC5">1</span><span style="color: #24292E">;</span></div><div class="shaku-underline shaku-underline-solid" style="left:6ch"><span class="shaku-underline-line" style="left:0ch">-----</span>This is line 1
+    .toEqual(`<pre class="shiki" style="color:#24292e;background-color:#fff"><div class="code-container"><code><div class="line "><span style="color: #D73A49">const</span><span style="color: #24292E"> </span><span style="color: #005CC5">a</span><span style="color: #24292E"> </span><span style="color: #D73A49">=</span><span style="color: #24292E"> </span><span style="color: #005CC5">1</span><span style="color: #24292E">;</span></div><div class="shaku-underline shaku-underline-solid" style="left:6ch"><span class="shaku-underline-line" style="left:0ch">-----</span>This is line 1
```
Comment (JSerZANP) [inline packages/remark-shaku-code-annotate/src/__tests__/index.test.ts:74]: same as above

File: packages/remark-shaku-code-annotate/src/index.ts [hunk 9]
```diff
@@ -165,11 +183,12 @@ export const remarkShakuCodeAnnotate = (
...
               isHighlightingBlock || shouldHighlighNextSourceLine;
             shouldHighlighNextSourceLine = false;
-
+            const shouldDim = isDimBlock || shouldDimNextSourceLine;
+            shouldDimNextSourceLine = false;
             const sourceLine = line.line;
             const prefix = `<div class="line${
               shouldHighlight ? " highlight" : ""
-            }">`;
+            } ${shouldDim ? "dim" : ""}">`;
```
Comment (JSerZANP) [inline packages/remark-shaku-code-annotate/src/index.ts:191]: this causes extra space for all lines if not dim

**Review by JSerZANP — CHANGES_REQUESTED** [review]
Nice PR,
just one trivial issue

File: examples/web/components/Playground.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (DongHY1) [reply examples/web/components/Playground.tsx]: > > but might it be slightly inconvenient but might it be slightly inconvenient但可能会有点不方便 但可能会有点不方便
> 
> Agreed, let's use `@dim` then.同意，那么让我们使用 `@dim` 。
> 
> I think we should separate the usage of these 3我认为我们应该将这 3 个的用法分开
> 
> 1. `@highlight` => to emphasize some lines `@highlight` => 强调某些行
> 2. `@dim` => to de-emphasize some lines `@dim` => 弱化某些行
> 3. `@focus` => emphasize some lines while de-emphasize the others, could be thought as a shorthand. `@focus` => 强调一些行，同时不强调其他行，可以被认为是一种速记。

Thanks for your advice. I will implement these.

File: packages/remark-shaku-code-annotate/src/index.ts (hunk 9 above)
Comment (DongHY1) [reply packages/remark-shaku-code-annotate/src/index.ts]: Thanks! The extra spaces have been removed.

File: packages/shaku-code-annotate/src/__tests__/index.test.ts [hunk 10]
```diff
@@ -141,7 +141,86 @@ describe("parseLine() can parse comment lines", () => {
...
+      config: {
+        mark: "below",
+      },
+    });
+  });
+  test("   @dim unexpected", async () => {
+    const func = () => parseLine("   @dim unexpected ");
+    expect(func).toThrow();
+  });
+  test("   @focus start", async () => {
```
Comment (JSerZANP) [inline packages/shaku-code-annotate/src/__tests__/index.test.ts:184]: You'd better keep PR small as possible, let's put @focus into another PR

Comment (DongHY1) [reply packages/shaku-code-annotate/src/__tests__/index.test.ts]: > You'd better keep PR small as possible, let's put `@focus` into another PR


Okay, I will open another PR.

**Comment by JSerZANP** [comment]
Great work, thanks!

**Pull request merged**